homeassistant/components/zoneminder/__init__.py

```diff
@@ -2,97 +2,169 @@
+import homeassistant.config_entries as config_entries
+from homeassistant.config_entries import ConfigEntry
 from homeassistant.const import (
     ATTR_ID,
     ATTR_NAME,
     CONF_HOST,
     CONF_PASSWORD,
     CONF_PATH,
+    CONF_PLATFORM,
+    CONF_SOURCE,
     CONF_SSL,
     CONF_USERNAME,
     CONF_VERIFY_SSL,
 )
-import homeassistant.helpers.config_validation as cv
-from homeassistant.helpers.discovery import async_load_platform
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers import config_validation as cv
+
+from . import const
+from .common import (
+    ClientAvailabilityResult,
+    async_test_client_availability,
+    create_client_from_config,
+    del_client_from_data,
+    get_client_from_data,
+    is_client_in_data,
+    set_client_to_data,
+    set_platform_configs,
+)
 
 _LOGGER = logging.getLogger(__name__)
-
-CONF_PATH_ZMS = "path_zms"
-
-DEFAULT_PATH = "/zm/"
-DEFAULT_PATH_ZMS = "/zm/cgi-bin/nph-zms"
-DEFAULT_SSL = False
-DEFAULT_TIMEOUT = 10
-DEFAULT_VERIFY_SSL = True
-DOMAIN = "zoneminder"
+PLATFORM_DOMAINS = tuple(
+    [BINARY_SENSOR_DOMAIN, CAMERA_DOMAIN, SENSOR_DOMAIN, SWITCH_DOMAIN]
+)
 
 HOST_CONFIG_SCHEMA = vol.Schema(
     {
         vol.Required(CONF_HOST): cv.string,
         vol.Optional(CONF_PASSWORD): cv.string,
-        vol.Optional(CONF_PATH, default=DEFAULT_PATH): cv.string,
-        vol.Optional(CONF_PATH_ZMS, default=DEFAULT_PATH_ZMS): cv.string,
-        vol.Optional(CONF_SSL, default=DEFAULT_SSL): cv.boolean,
+        vol.Optional(CONF_PATH, default=const.DEFAULT_PATH): cv.string,
+        vol.Optional(const.CONF_PATH_ZMS, default=const.DEFAULT_PATH_ZMS): cv.string,
+        vol.Optional(CONF_SSL, default=const.DEFAULT_SSL): cv.boolean,
         vol.Optional(CONF_USERNAME): cv.string,
-        vol.Optional(CONF_VERIFY_SSL, default=DEFAULT_VERIFY_SSL): cv.boolean,
+        vol.Optional(CONF_VERIFY_SSL, default=const.DEFAULT_VERIFY_SSL): cv.boolean,
     }
 )
 
-CONFIG_SCHEMA = vol.Schema(
-    {DOMAIN: vol.All(cv.ensure_list, [HOST_CONFIG_SCHEMA])}, extra=vol.ALLOW_EXTRA
+CONFIG_SCHEMA = vol.All(
+    cv.deprecated(const.DOMAIN, invalidation_version="0.118"),
+    vol.Schema(
+        {const.DOMAIN: vol.All(cv.ensure_list, [HOST_CONFIG_SCHEMA])},
+        extra=vol.ALLOW_EXTRA,
+    ),
 )
 
-SERVICE_SET_RUN_STATE = "set_run_state"
 SET_RUN_STATE_SCHEMA = vol.Schema(
     {vol.Required(ATTR_ID): cv.string, vol.Required(ATTR_NAME): cv.string}
 )
 
 
-def setup(hass, config):
+async def async_setup(hass: HomeAssistant, base_config: dict):
     """Set up the ZoneMinder component."""
 
-    hass.data[DOMAIN] = {}
+    # Collect the platform specific configs. It's necessary to collect these configs
+    # here instead of the platform's setup_platform function because the invocation order
+    # of setup_platform and async_setup_entry is not consistent.
+    set_platform_configs(
```

Comment:
We don't support mixing platform config and config entries. That should just be a migration. The platform config sections should be deprecated. What's the plan for the future?